### DISCUSSION TITLE
Building with CMake / FetchContent / ninja fails

### DISCUSSION THREAD

**Discussion by mgaimann** [BlueBrain/HighFive#745] [Questions] [answered]
Hi everyone,

I would like to use HighFive with CMake and FetchContent using:

```
FetchContent_Declare(
        highfive
        GIT_REPOSITORY https://github.com/BlueBrain/HighFive.git
        GIT_SHALLOW ON
        GIT_TAG v2.7.1
)
FetchContent_GetProperties(highfive)
FetchContent_MakeAvailable(highfive)
include_directories(BEFORE SYSTEM ${highfive_SOURCE_DIR} ${highfive_BINARY_DIR}/include)
```

I receive the following cryptic error message however:
```
CMake Error:
  Running

   '/home/user/.local/share/JetBrains/Toolbox/apps/CLion/ch-0/231.8770.66/bin/ninja/linux/x64/ninja' '-C' '/home/user/repos/myrepo/cmake-build-debug' '-t' 'recompact'

  failed with:

   ninja: error: build.ninja:1226: expected '=', got identifier
   -Wall -Wextra -Wno-unused-parameter -Werror -Wshadow -Wnon-virtual-dtor...
         ^ near here



CMake Generate step failed.  Build files cannot be regenerated correctly.

[Failed to reload]

```

I am not sure whether this is related to [this CMake issue](https://cmake.org/pipermail/cmake-developers/2012-October/017108.html) or [this question](https://gitlab.kitware.com/cmake/cmake/-/issues/21982).

Is there maybe a newline character `\n` somewhere in the CMakeLists.txt file of HighFive, which breaks generation with ninja?

The only places in the HighFive repo where the flags appear is [in the file CompilerFlagsHelpers.cmake](https://github.com/BlueBrain/HighFive/blob/794355c2bf83c01c9d5b2ce764cda38553d488e0/CMake/config/CompilerFlagsHelpers.cmake) as far as I can see.

Could the problem be that the flags are concatenated across lines which is interpreted as newlines and not as whitespace maybe?

Could you suggest a solution please? Ideally I would like to have all dependencies in my CMakeLists.txt file and build them automatically.
Thank you very much!

**Comment by 1uc** [comment]
I'm unable to reproduce this. In an otherwise empty directory put the following `CMakeLists.txt`
```
cmake_minimum_required(VERSION 3.26)
project(FetchItAll)

include(FetchContent)

FetchContent_Declare(
        highfive
        GIT_REPOSITORY https://github.com/BlueBrain/HighFive.git
        GIT_SHALLOW ON
        GIT_TAG v2.7.1
)
FetchContent_GetProperties(highfive)
FetchContent_MakeAvailable(highfive)
include_directories(BEFORE SYSTEM ${highfive_SOURCE_DIR} ${highfive_BINARY_DIR}/include)
```
and run
```
cmake -DCMAKE_GENERATOR=Ninja -B build-ninja . && cmake --build build-ninja
```
which builds the examples of HighFive without any complaints. Could you please check and describe how our tests differ?

**Reply by mgaimann** [reply]
Thank you! I carefully disabled and then enabled elements in my CMakeLists.txt and eventually managed to build HighFive. I traced the error back to the following line in my own CMakeLists.txt (which is strange since the error message referred to the flags of HighFive's CMakeLists.txt):

```
set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS} -g -Wall -Wextra -Wpedantic -Wdouble-promotion
        -Wformat=2 -Wformat-truncation -Wundef -fno-common -Wconversion -Wmisleading-indentation
") 
```

Apparently the line break here was messing up the building. I replaced it with the single line, and this seems to work now:
```
set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS} -g -Wall -Wextra -Wpedantic -Wdouble-promotion -Wformat=2 -Wformat-truncation -Wundef -fno-common -Wconversion -Wmisleading-indentation")
```

**Reply by 1uc** [reply]
Okay, since you confirmed that it's the culprit, we can change that. Since I've got somewhat definite plans to rewrite the HighFive CMake code, it would probably be good to have this issue reproducible so we can avoid it in the future.

I've tried with my version of CLion, but it works there. What version of CMake and ninja are you using?

**Reply by mgaimann** [reply]
I am using CMake version 3.24.2 and ninja version 1.10.2, hope that helps!

**Reply by 1uc** [reply, marked as answer]
There's a PR ready: #751 hopefully that solves the issue.
